src/clib/pioc_support.c

```diff
@@ -93,19 +94,57 @@ int PIOc_strerror(int pioerr, char *errmsg)
 int PIOc_set_log_level(int level)
 {
 #if PIO_ENABLE_LOGGING
-    char log_filename[NC_MAX_NAME];
-
     /* Set the log level. */
     pio_log_level = level;
+#endif /* PIO_ENABLE_LOGGING */
+    return PIO_NOERR;
+}
 
-    /* Create a filename with the rank in it. */
-    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
-    sprintf(log_filename, "pio_log_%d.txt", my_rank);
+/** Initialize logging
+ * Open log file, if not opened yet, or
+ * increment ref count if already open
+ */
+void pio_init_logging(void )
+{
+#if PIO_ENABLE_LOGGING
+    char log_filename[NC_MAX_NAME];
+
+    if (!LOG_FILE)
+    {
+        /* Create a filename with the rank in it. */
+        MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
+        sprintf(log_filename, "pio_log_%d.txt", my_rank);
+
+        /* Open a file for this rank to log messages. */
+        LOG_FILE = fopen(log_filename, "w");
+
+        pio_log_ref_cnt = 1;
+    }
+    else
+    {
+        pio_log_ref_cnt++;
```

Comment:
The other side of this problem (log files don't capture logs before pio_init() ) is that logs before a user sets pioc_set_log_level() is not captured by the current impl. With the proposed change , as before, all logs (including the logs before user sets pioc_set_log_level()) are directed to stdout.
I think it is cleaner to have init/finalize for the logging (and that these routines manage all the logic required to build and tear up logging support). The reference count is required in this implementation to keep track of inits/finalizes.
Setting log levels should be separate from setting up the logging infrastructure (log files etc).